target-gen/src/parser.rs

```diff
@@ -77,6 +77,13 @@ pub fn extract_flash_algo(
     let algorithm_binary = crate::algorithm_binary::AlgorithmBinary::new(&elf, buffer)?;
     algo.instructions = algorithm_binary.blob();
 
+    // Flash algorithms are all little endian. Swap the algorithm if necessary.
+    if !elf.little_endian {
+        for instruction in algo.instructions.chunks_exact_mut(4) {
```

Comment:
I wonder how this'll work with big endian Xtensas, which have 2 or 3 byte instructions. We don't support any such target yet, but could you place a TODO here mentioning <4 byte instruction sets?

Reply:
I think this will be fine, but I wouldn't mind seeing how it operates in practice. The issue is that the ELF file itself is all swapped, so we're effectively reading from a BE system here.

This already works for Thumb instructions, which are 2 bytes rather than 4. But still, I'll add the TODO.

Reply:
But if the elf contains the program in big endian, why do we have to flip it back? It should be loaded as is 🤔

Reply:
Because the program gets written with `.write_32()`, which swaps data around. Which is necessary to get things like the BRKPT instruction flipped.

If we don't flip the program from LE to BE, then we'll have to detect what state the core is in and flip the BRKPT instruction before it gets stuck on the front of the image. And we can't do that because the BRKPT instruction is prefixed well before the program is sent to the device.

Reply:
Another option is to note whether the flash image is BE or LE, and swap the instruction there. Then use `.write()` instead of `.write_32()` to avoid swapping data.

I can put together a patch that does that.

Reply:
> Because the program gets written with .write_32(), which swaps data around.

So the solution would be to not write the program with write_32, no?

Reply:
And to swap the contents `Self::algorithm_header()` if the target is BE, otherwise it'll execute an invalid instruction after returning rather than `bkpt`.

I'm testing a patch to do this.

Reply:
Yes, in my opinion the algorithm_header should be appropriate for the target, and probe-rs should not be responsible for transforming the program itself.

Reply:
I've updated it so that it now uses `.write()` and `.read()` and notes the endianness in a flag in the algorithm rather than unconditionally swapping it to be LE. This flag is then used to select a header to prefix onto the instructions.

The only change to `target_gen` now is to note the endianness of the ELF file which informs its decision as to which endianness to use for the header.